src/main/java/org/kohsuke/github/GitHub.java

```diff
@@ -482,7 +483,34 @@ void check(String apiUrl) throws IOException {
      * Otherwise this method throws {@link IOException} to indicate the problem.
      */
     public void checkApiUrlValidity() throws IOException {
-        retrieve().to("/", GHApiInfo.class).check(apiUrl);
+        try {
+            retrieve().to("/", GHApiInfo.class).check(apiUrl);
+        } catch (IOException ioe) {
+            if (isPrivateModeEnabled()) {
+                throw new IOException("GitHub Enterprise server (" + apiUrl + ") with private mode enabled");
+            }
+            throw ioe;
+        }
+    }
+
+    /**
+     * Ensures if a GitHub Enterprise server is configured in private mode.
+     *
+     * @return {@code true} if private mode is enabled. If it tries to use this method with GitHub, returns {@code
+     * false}.
+     */
+    private boolean isPrivateModeEnabled() {
+        try {
+            HttpURLConnection connect = getConnector().connect(getApiURL("/"));
+            if (connect.getResponseCode() == HttpURLConnection.HTTP_UNAUTHORIZED
+                    && connect.getHeaderField("Server") != null
+                    && connect.getHeaderField("Server").equals("GitHub.com")) {
```

Comment:
@cyrille-leclerc The goal of this PR was to provide a method to verify *if a URL represents a GitHub endpoint*, only that. A `HttpException 401 Unauthorized` does not involve that we have a valid GitHub endpoint.